From f7e07a0d858a3337d07d20eeb37bbbb93518388e Mon Sep 17 00:00:00 2001
From: Rheygine Medel <quic_rmedel@quicinc.com>
Date: Tue, 18 Apr 2023 09:06:33 -0700
Subject: [PATCH 2/5] SF: Add retry to EventThread postEvent

When consumer BitTube is under heavy Event load, like constantly
updating use case providing VSYNC events, it is possible for Hotplug
events to fail sending across the socket. In such case, retry
mechanism is needed to send it again. We would like not to invoke
retry logic so we are also doubling the BitTube size to avoid
Event backlog in the first place.

Change-Id: I7bcad9512cb6ecdac945fd765eb9258ccd88fee0
CRs-Fixed: 2700045
---
 .../surfaceflinger/Scheduler/EventThread.cpp  | 42 ++++++++++++-------
 1 file changed, 28 insertions(+), 14 deletions(-)

diff --git a/services/surfaceflinger/Scheduler/EventThread.cpp b/services/surfaceflinger/Scheduler/EventThread.cpp
index 96eccf290f..1b68fb8ead 100644
--- a/services/surfaceflinger/Scheduler/EventThread.cpp
+++ b/services/surfaceflinger/Scheduler/EventThread.cpp
@@ -196,7 +196,8 @@ EventThreadConnection::EventThreadConnection(EventThread* eventThread, uid_t cal
       : mOwnerUid(callingUid),
         mEventRegistration(eventRegistration),
         mEventThread(eventThread),
-        mChannel(gui::BitTube::DefaultSize) {}
+        mChannel(/* QTI_BEGIN */ gui::BitTube(
+                8 * 1024 /* default size is 4KB, double it */) /* QTI_END */) {}
 
 EventThreadConnection::~EventThreadConnection() {
     // do nothing here -- clean-up will happen automatically
@@ -700,6 +701,9 @@ void EventThread::generateFrameTimeline(VsyncEventData& outVsyncEventData, nsecs
 
 void EventThread::dispatchEvent(const DisplayEventReceiver::Event& event,
                                 const DisplayEventConsumers& consumers) {
+    /* QTI_BEGIN */
+    const uint8_t num_attempts = 3;
+    /* QTI_END */
     for (const auto& consumer : consumers) {
         DisplayEventReceiver::Event copy = event;
         if (event.header.type == DisplayEventReceiver::DISPLAY_EVENT_VSYNC) {
@@ -709,20 +713,30 @@ void EventThread::dispatchEvent(const DisplayEventReceiver::Event& event,
                                   event.vsync.vsyncData.preferredExpectedPresentationTime(),
                                   event.vsync.vsyncData.preferredDeadlineTimestamp());
         }
-        switch (consumer->postEvent(copy)) {
-            case NO_ERROR:
-                break;
-
-            case -EAGAIN:
-                // TODO: Try again if pipe is full.
-                ALOGW("Failed dispatching %s for %s", toString(event).c_str(),
-                      toString(*consumer).c_str());
-                break;
-
-            default:
-                // Treat EPIPE and other errors as fatal.
-                removeDisplayEventConnectionLocked(consumer);
+        /* QTI_BEGIN */
+        bool qtiNeedsRetry = true;
+        for (uint8_t attempt = 0; qtiNeedsRetry && (attempt < num_attempts); attempt++) {
+            /* QTI_END */
+            switch (consumer->postEvent(copy)) {
+                case NO_ERROR:
+                    /* QTI_BEGIN */ qtiNeedsRetry = false; /* QTI_END */
+                    break;
+
+                case -EAGAIN:
+                    // TODO: Try again if pipe is full.
+                    ALOGW("Failed dispatching %s for %s", toString(event).c_str(),
+                          toString(*consumer).c_str());
+                    /* QTI_BEGIN */ qtiNeedsRetry = true; /* QTI_END */
+                    break;
+
+                default:
+                    // Treat EPIPE and other errors as fatal.
+                    removeDisplayEventConnectionLocked(consumer);
+                    /* QTI_BEGIN */ qtiNeedsRetry = false; /* QTI_END */
+            }
+            /* QTI_BEGIN */
         }
+        /* QTI_END */
     }
     if (event.header.type == DisplayEventReceiver::DISPLAY_EVENT_VSYNC &&
         FlagManager::getInstance().vrr_config()) {
-- 
2.45.2

